### PR TITLE
KEYCLOAK-18749 AuthenticationSelectionResolver should only show elegible authenticators

### PR DESCRIPTION
**Pull request by thomasdarimont**
Previously AuthenticationSelectionResolver did not check if an authenticator
is actually configured for a user. We now honor the information from the configuredFor method
for authenticators that require a current user.

### CODE REVIEW

File: services/src/main/java/org/keycloak/authentication/AuthenticationSelectionResolver.java [hunk 1]
```diff
@@ -60,34 +64,58 @@
...
+            authenticationSelectionList.addAll(session.userCredentialManager()
+                    .getStoredCredentialsStream(realm, authenticatedUser)
+                    .filter(credential -> typeAuthExecMap.containsKey(credential.getType()))
+                    .map(CredentialModel::getType)
+                    .distinct()
+                    .map(credentialType -> new AuthenticationSelectionOption(session, typeAuthExecMap.get(credentialType)))
+                    .collect(Collectors.toList()));
+
+            //add authenticators that require a current user which are also configured for that user
+            for (AuthenticationExecutionModel exec : nonCredentialExecutions) {
```
Comment (mabartos) [inline services/src/main/java/org/keycloak/authentication/AuthenticationSelectionResolver.java:103]: If authenticated user is not null, this list of executions is fully iterated twice. In this case, what about to reduce these executions or completely avoid the second iteration over whole list? The condition for the `requiresUser` is evaluated twice. I know the list would change and potentially break the process of executing actions for the `nonCredentialExecutions` list in the future, but at this moment, there's no other action for the list. Is there any deeper intention why to use this approach?

**Review by mabartos — COMMENTED** [review]
@thomasdarimont Thanks for the fix! In general, it LGTM, but I have just question/comment below. Moreover, could you please add any test case for this in order to avoid the similar situation with the bug in the future?

EDIT: It seems the GH actions failed due to this PR.

**Review by mposolda — CHANGES_REQUESTED** [review]
@thomasdarimont Thanks for the PR.

I am probably missing some context and the concrete use-case, but from what I remember the "nonCredential" executions contains the authenticators, which are not explicitly configured for the user, are expected to be shown in the alternatives. So for example when there is subflow like this with 3 alternatives:
- WebAuthn ALTERNATIVE
- OTP ALTERNATIVE
- some-non-credential-auth ALTERNATIVE

And the user has OTP, but doesn't have WebAuthn, then I think it is currently expected, that he would see this:
- OTP
- some-non-credential-auth

In the authentication selection screen.

Is there any concrete use-case for this issue? The JIRA does not seem to contain any concrete use-case. If it is possible to add some use-case to it, it will be nice.

Finally, there are some failing test as @mabartos mentioned. Which I guess is exactly due the fact that existing implementation expects the behaviour above.

Hopefully we can come with some improvement to address both the current use-cases and the new one.

**Comment by thomasdarimont** [comment]
Hi @mposolda, 

I think the context for this PR is this discussion on the keycloak-dev mailing list:
https://groups.google.com/g/keycloak-dev/c/MqpYd7kKa3U/m/0RJcCLdCBQAJ

I'll take a look at the remarks later and see what I can do to fix the failing tests.

**Comment by mposolda** [comment]
@thomasdarimont Is this PR still valid? Do you have a chance to check existing comments?

**Comment by mposolda** [comment]
@thomasdarimont Closing this PR for now. Feel free to re-open once you have a chance to look at the comments above

**Pull request closed without merging**